sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,743 @@
+# Explicit SIMD Programming Extension for DPC++
+
+## Introduction
+
+The main motivation for introducing the Explicit SIMD Programming (ESP) DPC++
+extension is enabling low-level efficient programming for Intel Gen
+architectures. More specifically, Explicit SIMD provides the following
+additional features:
+- Manual vectorization of device code using the `simd` class mapped to Gen's
+  general register file. This allows to write efficient code not relying on
+  further widening by the compiler, as with traditional SPMD programming.
+- Low-level APIs efficiently mapped to Gen architecture, such as block reads.
+
+Here are some future directions in which this API is intended to evolve:
+- enabling this extension for other architectures, such as x86, with extracting
+  and clearly marking generic and target-dependent API portions
+- aligning with `std::simd` and maybe providing `std::simd` implemenation atop
+  `sycl::intel::gpu`
+
+## Explicit SIMD execution model
+
+Explicit SIMD execution model is a basically an equivalent of the base SYCL
+execution model with subgroup size restricted to 1. Which means each subgroup
+maps to a single hardware thread. All standard SYCL APIs continue to work,
+including `sycl::intel::sub_group` ones, which become either a no-op or
+trivial. E.g. a barrier becomes just a memory fence for a compiler, collectives
+just return the value in the single work-item. Another consequences of the unit
+subgroup size is guaranteed independent forward progress between work-items on
+some Gen architecture generations.
+
+## Explicit SIMD extension APIs
+
+Explicit SIMD APIs can be used only in code to be executed on Intel Gen
+architecture devices and the host device for now. Attempt to run such code on
+other devices will result in error. 
+
+Kernels and `SYCL_EXTERNAL` functions using ESP must be explicitly marked with
+the `[[sycl_explicit_simd]]` attribute. Subgroup size query within such
```

Comment:
I don't want to bikeshed on this, but just a note: I think `[[sycl_explicit_simd]]` is better than a macro (for all the reasons @rolandschulz pointed out) but the spelling here is different to attributes like `[[intel::reqd_work_group_size]]`.

We should decide what rules we're following for attributes (if any).  I'll open an issue for this as well.

Reply:
I think we should also update all the examples to align with this proposal: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/KernelRHSAttributes/SYCL_INTEL_attribute_style.asciidoc

Reply:
Done. (@bader - I assume alignment means function type attribute placed before definition instead of function attribute placed before the declaration)